PowerPC fixes

kernigh opened the pull request:
This pull request includes commits by @davidgiven (dtrg-experimental-powerpc) and by me (kernigh-linuxppc). These fixes improve our PowerPC back end from broken to mostly working. There are still problems, but we can now build ELF binaries for Linux.

<s>These commits don't enable the build. People who want to build this need to add "linuxppc" to `vars.plats` in the root build.lua.</s> Actually, 45a9505 does add "linuxppc" to the build.
- 80cb6ba removes the broken RELOH2 from led, and extends RELOPPC to handle addis/ori pairs, using a special pseudo-opcode `la`.
- 5b69777 renames it from `la` to `li32`.
- e4ce7da edits examples/hilo.p for big-endian platforms.
- 9ec2918 increases MAXREGS in ncgg from 80 to 200.
- 65c2a8a removes the unused stackadjust/stackoffset() from ncg.
- e2ccc8f fixes a major problem with powerpc/ncg by adding "kills" parts to "sti" rules.

Most of the other commits fix or improve the code generation rules in mach/powerpc/ncg/table or their helpers in mach/powerpc/libem

Most of the example programs are working. Basic is broken. (You can run `hilo_b.linuxppc` if you have an empty file named `hilo.d` in the current directory, but the correct guess is always zero.) In startrek, I can navigate the Enterprise and fire phasers but can't dock at a starbase.

This pull request doesn't include my Mac OS X stuff. That's in a different branch (https://github.com/kernigh/ack/tree/kernigh-osx) and builds Mach-o binaries for both PowerPC and Intel.

davidgiven committed:
Readjust after merge to default.

--HG--
branch : dtrg-experimental-powerpc

davidgiven committed:
Rename branch.

--HG--
branch : dtrg-experimental-powerpc-branch

davidgiven committed:
Merge from default.

--HG--
branch : dtrg-experimental-powerpc-branch

davidgiven committed:
Mostly add support for the experimental and largely broken linuxppc platform.
(Doesn't quite build.)

--HG--
branch : dtrg-experimental-powerpc-branch

davidgiven committed:
Eliminate the RELOH2 relocation, as it never worked --- the address would be
calculated incorrectly because of overflow errors.

Replace it with an extended RELOPPC relocation which understands addis/ori
pairs; add an la pseudoop to the assembler which generates these and the
appropriate relocation. Make good.

--HG--
branch : dtrg-experimental-powerpc-branch

kernigh committed:
Merge branch 'default' into kernigh-linuxppc

kernigh committed:
Add the missing .lar4 and .sar4 for powerpc.

Inspired by the sparc code (mach/sparc/libem/lar.s).  My powerpc code
might still have bugs, but it's enough for examples/hilo.mod to work.

May need to 'make clean' or touch a build.lua file, so ackbuilder can
notice the new lar4.s and sar4.s files and build them.

kernigh committed:
Fix hilo.p for big-endian platforms.

Unless it is packed, a Pascal char is a C int.  Using C types, hilo.p
passed an int *buf to uread(), which expected a char *buf.  Then
uread() wrote the char on the end of the int.  This worked on
little-endian platforms.  This failed on big-endian platforms, as
writing the value to the big end of an int multiplied it by 16777216.

The fix is to use a packed array [0..0] of char in Pascal.  I also
change 'string' to a packed array, though this is not a necessary part
of the fix.

kernigh committed:
In ncgg, increase MAXREGS from 80 to 200.

I need this to add more registers to powerpc.

kernigh committed:
Enable the Hall check again, and get powerpc to pass it.

Upon enabling the check, mach/powerpc/ncg/table fails to build as ncgg
gives many errors of "Previous rule impossible on empty stack".  David
Given reported this problem in 2013:
  https://sourceforge.net/p/tack/mailman/message/30814694/

Commit c93cb69 commented out the error in util/ncgg/cgg.y to disable
the Hall check.  This commit enables it again.  In ncgg, the Hall
check is checking that a rule is possible with an empty fake stack.
It would be possible if ncg can coerce the values from the real stack
to the fake stack.  The powerpc table defined coercions from STACK to
{FS, %a} and {FD, %a}, but the Hall check didn't understand the
coercions and rejected each rule "with FS" or "with FD".

This commit removes the FS and FD tokens and adds a new group of FSREG
registers for single-precision floats, while keeping FREG registers
for double precision.  The registers overlap, with each FSREG
containing one FREG, because it is the same register in PowerPC
hardware.  FS tokens become FSREG registers and FD tokens become FREG
registers.  The Hall check understands the coercions from STACK to
FSREG and FREG.  The idea to define separate but overlapping registers
comes from the PDP-11 table (mach/pdp/ncg/table).

This commit also removes F0 from the FREG group.  This is my attempt
to keep F0 off the fake stack, because one of the stacking rules uses
F0 as a scratch register (FSCRATCH).

kernigh committed:
Rename our pseudo-opcode 'la' to 'li32'.

GNU as has "la %r4,8(%r3)" as an alias for "addi %r4,%r3,8", meaning
to load the address of the thing at 8(%r3).  Our 'la', now 'li32',
makes an addis/ori pair to load an immediate 32-bit value.  For
example, "li32 r4,23456789" loads a big number.

kernigh committed:
Multiple tweaks to plat/linuxppc/descr

Don't define __POWERPC.  I don't know any other compiler that defines
__POWERPC and don't want to invent a new macro.  Apple's gcc 4.0.1
from Xcode 2.5 defines __ppc__, _ARCH_PPC, __POWERPC__.  Debian's gcc
4.9.2-10 defines _ARCH_PPC, __PPC__, __powerpc__, __PPC, __powerpc,
PPC, powerpc.

Move the base vm address from 0x80000000 down to 0x10000000, as this
is where Debian loads /bin/true.  This is still higher than the base
addresses for linux386 and linux68k.

Sync led's arguments with linux386.

kernigh committed:
Remove the "invalid" stacking rule.

When ncg fell back on this rule, it did emit the string "invalid" in
the assembly code and caused a syntax error in the assembler.

Adjust the stacking rules so we can stack LOCAL, CONST, and LABEL
without falling back on the "invalid" rule, and so we can stack them
when we have no free register except the scratch register.

kernigh committed:
For "pat ass", move fake stack to real stack before adjusting SP.

This fixes code that tried to "addi SP, SP, 4" to drop a value that
was in a register, not on the real stack.

Add a rule to optimize "asp 4" (which becomes "loc 4" "ass") when
the value being dropped is already in a GPR.

kernigh committed:
Rewrite fef 8 in powerpc assembly.

In EM, fef splits a float into exponent and fraction.  The old C code,
given an infinite float, got stuck in an infinite loop.  The new
assembly code doesn't loop; it extracts the IEEE exponent.

kernigh committed:
Add a rule for sdl ldl $1==$2 to work around a bug.

In our powerpc table, sdl fails to kill the old value of the local.
This is a bug, because a later ldl can load the old value instead of
the newly stored value.  By rewriting "sdl 0" "ldl 0" as "dup 8" "sdl
0", the newly added rule works around the bug, but only when the ldl
is immediately after the sdl.

This rule improves code that uses double-precision floating point.
The output of printf("%f", 6.0) in C changes from all zero digits to
"6000000" but still doesn't print the decimal point.  The result of
atof("-123.456") becomes correct.  In startrek, I can now move the
Enterprise, but I still can't fire phasers without crashing the game.

We already have a rule for stl lol $1==$2.  We had two copies of the
rule, so I am deleting the second copy.

kernigh committed:
Define the begdata, begrom, begbss symbols for linuxppc.

I copied the definitions from linux386 and linux68k.

This change also moves _errno and the other common symbols in boot.s
from .text to .bss.  Common symbols belong in .bss, but the assembler
seems dumb enough to put them in any section.

kernigh committed:
Remove .linenumber and .filename; use hol0 and hol0+4.

We need this because some .e files in lang/ are using 'loe 0' and 'lae
4' to load the line number from hol0 and filename from hol0+4.

kernigh committed:
Merge branch 'default' into kernigh-linuxppc

kernigh committed:
Rename SCRATCH to RSCRATCH.  Never stack RSCRATCH nor FSCRATCH.

Rename the scratch gpr (currently r11) from SCRATCH to RSCRATCH so I
can search for RSCRATCH without finding FSCRATCH.  I also want to
avoid confusion with the SCRATCH keyword of the old code generator (cg
which came before ncg).

Change the stacking rules to prevent stacking of RSCRATCH or FSCRATCH
or any other GPR or FPR that isn't an allocatable REG or FREG.  Then
ncgg rejects any rule that tries to stack a GPR or FPR, so change such
rules to stack a REG or FREG.

kernigh committed:
Remove most of GPRE from mach/powerpc/ncg/table

We only need GPRE in a few places where we write {GPRE, regvar(...)}
because ncgg can't parse plain regvar(...).  In all other places, a
plain GPR works.

Also remove gpr_gpr_gpr and a few other unused and fake instructions
from the list of instructions.

kernigh committed:
Remove stackadjust and stackoffset() from ncg.

This feature has never been used since its introduction, more than 3
years ago, in David Given's commit c93cb69 of May 8, 2013.  The commit
was for "PowerPC and M68K work".  I am not undoing the entire commit.
I am only removing the stackadjust and stackoffset() feature.

This commit removes the feature from my branch kernigh-linuxppc.  This
removal includes the mach/proto/ncg parts.  The default branch already
removed most of the feature, but kept the mach/proto/ncg parts.  That
removal happened in commit 81778b6 of May 13, 2013 (which was a merge;
git diff af0dede 81778b6).  The branch dtrg-experimental-powerpc
merged the default branch but without the removal.  That merge was
commit 4703db0f of Sep 15, 2016 (git diff 8c94b13 4703db0).  My branch
kernigh-linuxppc is off branch dtrg-experimental-powerpc, so I can no
longer get the removal by merging default.

David Given described the stackadjust feature in
  https://sourceforge.net/p/tack/mailman/message/30814691/

The instruction stackadjust would add a value to the offset, and the
function stackoffset() would return this offset.  One would use this
to track sp - fp, then omit the frame pointer by not keeping fp in a
register.

kernigh committed:
In powerpc table, fix macros los() and his().

Change the operator in his() from a - minus to a + plus.  When los(n)
becomes negative, then his(n) needs to add 0x10000, not subtract it.

Also change los(n) to do the sign extension, because smalls(los(n))
should be true, not false.

Also change hi(n) and lo(n) to wrap n in parentheses, as (n), because
these are macros and n might still contain operators.

kernigh committed:
Remove unused parts of mach/powerpc/ncg/table

Remove unused tokens GPRINDIRECTLO, HILABEL, LOLABEL, LABELI.  Also
remove an #if 0 ... #endif group of patterns.

kernigh committed:
Refactor how powerpc ncg pushes constants.

When loc (load constant) pushes a constant, it now checks the value of
the constant and pushes any of 7 tokens.  These tokens allow stack
patterns to recognize 16-bit signed integers (CONST2), 16-bit unsigned
integers (UCONST2), multiples of 0x10000 (CONST_HZ), and other
interesting forms of constants.

Use the new constant tokens in the rules for adi, sbi, and, ior, xor.
Adjust a few other rules to understand the new tokens.

Require that SUM_RC has a signed 16-bit constant, and OR_RC and XOR_RC
each have an unsigned 16-bit constant.  The moves from SUM_RC, OR_RC,
XOR_RC to GPR no longer touch the scratch register, because the
constant is not too big.

kernigh committed:
Simplify moves to and from IND_RC_*

Now that SUM_RC always has a signed 16-bit constant, it happens that
the various IND_RC_* tokens also have a signed 16-bit constant, so
we no longer need to touch the scratch register.

kernigh committed:
Remove IND_LABEL_W and IND_LABEL_D

Because li32 always loads a label into a GPR, it is sufficient to
coerce LABEL to REG, then use IND_RC_W or IND_RC_D for indirection
through the label.

kernigh committed:
Add "kills MEMORY" to powerpc sti rules.

Adjust some of the loi rules (and associated moves) so we can identify
the tokens that must be in MEMORY.

With this commit, I can navigate the Enterprise even if I comment out
my work-around from e22c888.

kernigh committed:
Rewrite .fif8 to avoid powerpc64 fctid

This fixes the SIGILL (illegal instruction) in startrek when firing
phasers.  The 32-bit processors in my PowerPC Mac and in QEMU don't
have fctid, a 64-bit instruction.

I got the idea from mach/proto/fp/fif8.c to extract the exponent,
clear some bits to get an integer, then subtract the integer from
the original value to get the fraction.

kernigh committed:
Add costs to powerpc instructions.

Also show how andi., andis., or., set condition codes.

kernigh committed:
In powerpc ncg, add a speed hack for sti 8.

ncg is too slow with this many registers.  A stack pattern "with GPR
GPR GPR" or "with REG REG REG" takes too long to pick registers,
causing ncg 8 to take about 2 seconds on each sti 8.  I introduce
REG_PAIR and there are only 4 such pairs.

For programs that use sti 8 (including C programs that copy 8-byte
structs), this speed hack improves the ncg run from several seconds to
almost instantaneous.

Also add a few COMMENT(...) lines in stacking rules.

davidgiven commented:
Thank you very much for that! That's awesome, and I'm so sorry that you now understand how ncg works... The detailed commit messages are strongly appreciated. (Particularly thanks for removing the stackadjust stuff, which I never got round to doing; that was a failed experiment.)

Before I merge it: I did a test build, and it looks like the PowerPC ncg is taking an age to build anything. Building mandelbrot.c takes over a second, all of which is in ncg:

```
$ ACKDIR=/tmp/ack-build/staging /tmp/ack-build/staging/bin/ack -mlinuxppc -c -o /tmp/ack-build/obj/examples/mandelbrot_c_linuxppc/main/mandelbrot/mandelbrot.o examples/mandelbrot.c  -O6 -t
$ time /tmp/ack-build/staging/lib/ack/linuxppc/ncg mandelbrot.g > /dev/null
real    0m1.103s
user    0m1.100s
sys 0m0.000s
```

If you run ncg with -d1 it dumps details about the codegen; I think it's hitting finding a pathalogical case while doing instruction selection, possibly because it can't decide between different code paths; I've seen this before, but unfortunately can't remember the details.

It shouldn't be taking nearly that long. It's definitely related to the register file, because reducing the number of registers speeds things up. My gut feeling is something to do with REG_PAIR or FSREG/FREG is causing a combinatorial explosion. Could you take a look, please?

kernigh committed:
Remove f14 to f31 from FREG and FSREG.

This would have happened later, if f14 to f31 became regvar (like r13
to r31 are now).  I am doing it now because ncg is too slow for rules
"with FREG FREG uses FREG".  We use such rules for adf 8 and other EM
instructions that operate on 2 floats.  Like my last commit cfbc537,
this commit speeds ncg by removing choices for register allocation.

kernigh commented:
It was FREG. I have added 99dee0a to reduce the number of registers; I don't know how to do anything else. One of the problem rules was

```
   pat adf $1==INT64                  /* Add double */
            with FREG FREG
                    uses FREG
                    gen
                            fadd %a, %2, %1
                    yields %a
```

ncg stalled before deciding to use `fadd f31,f2,f1`.

davidgiven commented:
It's still noticeably chuggy when it builds the PowerPC stuff, but... you know what, that's fine. We can see if this is fixable later. I'll merge. Thank you so much for your work!
